src/ClusterProviders/Proto.Cluster.Consul/ConsulProvider.cs

```diff
@@ -0,0 +1,134 @@
+﻿// -----------------------------------------------------------------------
+//   <copyright file="ConsulProvider.cs" company="Asynkron HB">
+//       Copyright (C) 2015-2017 Asynkron HB All rights reserved
+//   </copyright>
+// -----------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Consul;
+
+namespace Proto.Cluster.Consul
+{
+    public class ConsulProvider : IClusterProvider
+    {
+        private TimeSpan _blockingWaitTime;
+        private ConsulClient _client;
+        private string _clusterName;
+        private TimeSpan _deregisterCritical;
+        private string _id;
+        private ulong _index;
+        private TimeSpan _refreshTtl;
+        private bool _shutdown;
+        private TimeSpan _ttl;
+
+
+        public async Task RegisterMemberAsync(string clusterName, string host, int port, string[] kinds)
+        {
+            _id = $"{clusterName}@{host}:{port}";
+            _clusterName = clusterName;
+            _index = 0;
+            _ttl = TimeSpan.FromSeconds(3);
+            _refreshTtl = TimeSpan.FromSeconds(1);
+            _deregisterCritical= TimeSpan.FromSeconds(10);
+            _blockingWaitTime = TimeSpan.FromSeconds(20);
+
+            _client = new ConsulClient(c =>
+            {
+               //TODO: config
+            });
+
+            var s = new AgentServiceRegistration
+            {
+                ID = _id,
+                Name = clusterName,
+                Tags = kinds,
+                Address = host,
+                Port = port,
+                Check = new AgentServiceCheck
+                        {
+                            DeregisterCriticalServiceAfter = _deregisterCritical,
+                            TTL = _ttl
+                        }
+            };
+            await _client.Agent.ServiceRegister(s);
+
+
+            //register a unique ID for the current process
+            //similar to UID for Akka ActorSystem
+            //TODO: Orleans just use an int32 for the unique id called Generation.
+            var kvKey = $"{_clusterName}/{host}:{port}"; //slash should be present

```

Comment:
Will do, currently we just use some semi random/unique string. but this could be something simpler like a long value with the current tick.
But I will address the comment in the code